### DISCUSSION TITLE
Periodic BC in FV

### DISCUSSION THREAD

**Discussion by wehrad** [idaholab/moose#29537] [Q&A Modules: General]
Hi everyone,

I'm working on an INSFV app and would need to apply periodic boundary conditions to test it on analytical solutions. I can't seem to find any example of the Periodic Action used in FV, and when I'm trying to add it in `[BCs]` it's simply being disregarded, (which was somehow expected).

The geometry is a simple 2D block so periodicity should be easy to catch.

Here is my input file: https://github.com/AdrienWehrle/diuca/blob/PR_ViscousFlowFVTests/inputs/viscous_diuca/iceslab_fv_2d_SI_nv.i

I'm on moose-libmesh 2024.12.02 and moose-tools 2024.11.13, and the latest version of moose on master.

Thanks a lot in advance for your help!

**Comment by GiudGiud** [comment]
Hello

We have only implemented periodic BCs for nodal variables.
For FV, it would be possible to do it through the Functor system:
- create a new object deriving from the FVDirichletBC class
- compute the geometrical transformation from the local face to the face where the value should be the same
- create a functor argument (FaceArg most likely) for that other face
- return the variable_functor(face arg, current state)

This would connect the value on one face to another

**Reply by wehrad** [reply]
Should I base that new object on https://github.com/idaholab/moose/blob/cf0cd9d67ad6180c403df50ab5cfc220c9bfa6df/framework/src/actions/AddPeriodicBCAction.C ?

**Reply by wehrad** [reply]
While I understand your points 1, 3 and 4, I'm not sure I understand what you mean by geometrical transformation from the local face to the other. In principle I do, but in MOOSE terms, I don't really see how that would be implemented.

**Reply by GiudGiud** [reply]
no you should base it off of FVDirichletBC

**Reply by GiudGiud** [reply]
actually FVFunctorDirichletBC.
I ll take a look today. This wont work in parallel for now

**Reply by wehrad** [reply]
I see! Thanks a lot!

**Reply by GiudGiud** [reply]
can you please try to use this
https://github.com/idaholab/moose/pull/29540
you will need to define the geometric transformation between the two boundaries, create a new functor that is evaluated on the "other boundary" when called on the "first" boundary

use that functor in a FVDirichletBC

**Reply by wehrad** [reply]
Thanks for working on this! I'm trying to follow the steps but struggle a bit. I wrote the geometric transformations (very simple in my case since it's just a 2D block with regular mesh):

```
[Functions]
  [transform_x]
    type = ParsedFunction
    expression = 'length'
    symbol_names = 'length'
    symbol_values = '${length}'
  []
  [transform_y]
    type = ParsedFunction
    expression = '1'
  []
[]
```

then I passed them in the new object you created but I'm not sure that's actually right.

```
[FunctorMaterials]
  [ice]
    type = FVIceMaterialSI
    velocity_x = "vel_x"
    velocity_y = "vel_y"
    pressure = "pressure"
    output_properties = 'mu_ice rho_ice eps_xx eps_yy sig_xx sig_yy eps_xy sig_xy'
    outputs = "out"
  []
  [translate_vel_x]
    type = ADFunctorTransformFunctorMaterial
    prop_names = 'transformed_vel_x'
    prop_values = 'vel_x'
    x_functor = 'transform_x'
    y_functor = 'transform_y'
    z_functor = '0'
  []
  [translate_vel_y]
    type = ADFunctorTransformFunctorMaterial
    prop_names = 'transformed_vel_y'
    prop_values = 'vel_y'
    x_functor = 'transform_x'
    y_functor = 'transform_y'
    z_functor = '0'
  []
  [translate_pressure]
    type = ADFunctorTransformFunctorMaterial
    prop_names = 'transformed_pressure'
    prop_values = 'pressure'
    x_functor = 'transform_x'
    y_functor = 'transform_y'
    z_functor = '0'
  []
[]
```

If by any chance those objects are correct, then I think should pass them in a `DirichletBC` as you said. I understand it as the following, where `translated_vel_y` coming from the `left` boundary would be applied through a `DirichletBC` on the `right` boundary. In that sense I feel like a `CoupledVarDirichletBC` would be needed, like the one I worked on for FE but for FV, [here](https://github.com/AdrienWehrle/diuca/blob/main/src/bcs/CoupledVarDirichletBC.C). Then it would give a BC like the following:
```
[periodic_right]
    type = CoupledVarDirichletBC
    variable = vel_x
    boundary = 'right'
    v = transformed_vel_x
[]
```

But that feels very odd compared to the steps you gave above which makes me think I might be off track here.

My input file is [here](https://github.com/AdrienWehrle/diuca/blob/PR_ViscousFlowFVTests/inputs/viscous_diuca/iceslab_fv_2d_SI_nv.i)

**Reply by GiudGiud** [reply]
Use a (AD?)FunctorDirichletBC.

We have not created another variable here, just a functor material property

**Reply by wehrad** [reply]
Ah yes indeed, missed that!

Now I have

```
[periodic_vel_x]
    type = FVFunctorDirichletBC
    variable = vel_x
    boundary = 'right'
    functor = transformed_vel_x
[]
[periodic_vel_y]
    type = FVFunctorDirichletBC
    variable = vel_y
    boundary = 'right'
    functor = transformed_vel_y
[]
[periodic_pressure]
    type = FVFunctorDirichletBC
    variable = pressure
    boundary = 'right'
    functor = transformed_pressure
[]
  ```

and get 

```
The AD object 'periodic_vel_x' is requesting the functor 'transformed_vel_x' as a non-AD functor even though it is truly an AD functor, which is not allowed, since this may unintentionally drop derivatives.
```

`ADFunctorDirichletBC` doesn't exist afaik, should I create it?

**Reply by GiudGiud** [reply]
it does exist as `FVADFunctorDirichletBC`

**Reply by wehrad** [reply]
Oh now I found it browsing through the source code on Github but I was initially checking the [source documentation](https://mooseframework.inl.gov/source/index.html) where it doesn't seem like it appears. Thanks!

**Reply by wehrad** [reply]
For now I get a segmentation fault on [this](https://github.com/AdrienWehrle/diuca/blob/PR_ViscousFlowFVTests/inputs/viscous_diuca/iceslab_fv_2d_SI_nv.i) input file that includes your new object, which is probably due to a failing transform I wrote. And somehow the `dbg` method and the `devel` one can't compile successfully for your branch on my side for now, working on it. But I think that input file is pretty close to the final version.

**Reply by wehrad** [reply]
In the meantime, do you feel like your new object is used accordingly in that input file ([here](https://github.com/AdrienWehrle/diuca/blob/PR_ViscousFlowFVTests/inputs/viscous_diuca/iceslab_fv_2d_SI_nv.i))? Thanks a lot!

**Reply by GiudGiud** [reply]
Hello

I think you mean to translate by a negative amount to go from right to left rather than a positive amount

**Reply by wehrad** [reply]
ah yes indeed that was it, sorry about that! Thanks!

Now I'm only using this x transform for each of the three variables (`vel_x`, `vel_y` and `pressure`) but it seems like the boundary is not handled well as you can see in the screenshot below, showing velocity magnitude.

![image](https://github.com/user-attachments/assets/339c3496-6b89-4878-9079-da802f59ef18)

Created with [this](https://github.com/AdrienWehrle/diuca/blob/PR_periodic_FV/inputs/viscous_diuca/iceslab_fv_2d_SI_nv.i) input file. Any idea? It feels like the transform is correct now. I'll keep investigating!

**Reply by GiudGiud** [reply]
Velocity magnitude does not give the full picture. Can you plot all 3 variables to see what could have gone wrong?
Which ones are supposed to be periodic?

**Reply by wehrad** [reply]
>  Which ones are supposed to be periodic?

All three should be periodic. Setup is an inclined infinite block of ice (inclination given by a projection of gravity rather than a slope on the mesh).

Here are the three variables ('pressure', 'vel_x', 'vel_y'):

![image](https://github.com/user-attachments/assets/2c1831ca-7cae-496a-a664-504070b0e7a0)

![image](https://github.com/user-attachments/assets/fcf9777b-00f0-4a2f-b031-b987331fa328)

![image](https://github.com/user-attachments/assets/79cfa3a7-6892-4236-9874-889468418476)

**Reply by GiudGiud** [reply]
Why would pressure be periodic? Should there not be an offset from the hydrostatic pressure from the incline ? (Which is easy to add here to the functor)

**Reply by wehrad** [reply]
> Why would pressure be periodic?

I had it periodic because that's what I set in the FE test too and it was working fine (this setup is tested against an analytical solution). But I suppose periodic velocities in x and y pretty much results in a periodic pressure anyway.

> Should there not be an offset from the hydrostatic pressure from the incline ?

I'm not sure there should be one since the coordinate system is rotated on the inclined gravity? At least in FE again, I didn't apply any offset and the pressure was as expected (i.e., a vertical gradient from the overburden pressure that was the same all along the block)

**Reply by GiudGiud** [reply]
If the offset can be precompiled from just the height difference, I recommend you use that instead of a periodic

A wrong pressure will definitely break the velocity components too

**Reply by wehrad** [reply]
what I struggle understanding is that there seems to be a rotation needed in the pressure in the FV system (on this one I removed the pressure periodicity hence why there's no issue at the boundary): 
![image](https://github.com/user-attachments/assets/6f4592dd-c947-43b8-90d9-0476d6b84afa)

While I did pretty much the same as for FE, i.e. projecting gravity onto the slope of the slab to keep it strictly horizontal. Yet in FE I had no such issue with pressure projection (with or without periodic pressure is the same), see below the pressure profile with depth being constant along x:
![image](https://github.com/user-attachments/assets/0e3a06d4-d962-49c5-bd43-0c1c4708753e)

NOTE: both slabs are perfectly horizontal in the X-Y plane, and slab slope is applied with a projection on gravity:
```SWIG
# change coordinate system to add a slope
gravity_x = '${fparse sin(bed_slope / 180 * pi) * 9.81 }'
gravity_y = '${fparse - cos(bed_slope / 180 * pi) * 9.81}'
```

**Reply by GiudGiud** [reply]
The direction of gravity does not look the same in both plots.

Is the pressure profile essentially coming mostly from gravity? Or is it due to flow? Do you get a flat pressure without gravity ?



If the mesh itself is already slanted, why do you need to have a gravity vector that is not aligned with the axis?

Do you get the same results when rotating the mesh using a transformGenerator in the Mesh block and a Y-only gravity vector

**Reply by wehrad** [reply]
> The direction of gravity does not look the same in both plots.

Right, although gravity is applied exactly the same way in the FE and FV problems. That's where I'm puzzled.

For both FE and FV setups I use:
```SWIG
# change coordinate system to add a slope
gravity_x = '${fparse sin(bed_slope / 180 * pi) * 9.81 }'
gravity_y = '${fparse - cos(bed_slope / 180 * pi) * 9.81}'
```

and then for FV (on `vel_x`):
```SWIG
  [u_gravity]
    type = INSFVMomentumGravity
    variable = vel_x
    momentum_component = 'x'
    rho = ${rho}
    gravity = '${gravity_x} ${gravity_y} 0.'
  []
```
and for FE (on the `velocity` vector):
```SWIG
  [gravity]
    type = INSADGravityForce
    variable = velocity
    gravity = '${gravity_x} ${gravity_y} 0.'
  []
```

> Is the pressure profile essentially coming mostly from gravity? Or is it due to flow? Do you get a flat pressure without gravity ?

There should be no other pressure variation than originating from the flow originating from gravity. It's simply a viscous material deforming due to gravity. I get a constant null pressure without gravity as expected, since all of it comes from gravity.

> If the mesh itself is already slanted, why do you need to have a gravity vector that is not aligned with the axis?

The mesh is not slanted, the mesh itself is horizontal at all times, and that both in the FE and FV setups. I decided to slant gravity instead of the mesh itself to make the analysis easier, instead of slanting the mesh, and then projecting the velocity on the slope (which is what I need for the analytical solution). It worked just fine with FE and gave the right results, that's why I'm puzzled.

Here is basically what I'm trying to do with FV (after I managed getting the right results in FE): 

![image](https://github.com/user-attachments/assets/ab092d7f-f0c6-4602-9597-1a0ecd779cf0)

And I need to evaluate the velocity in the x,z dimensions (on the schematics).

> Do you get the same results when rotating the mesh using a transformGenerator in the Mesh block and a Y-only gravity vector

I was working on that before you wrote, and I hence tried with the objects I found first (that is, the `FillBetweenCurvesGenerator` combined with `ParsedGenerateSideset` and it worked great although probably overshot here). I get a similar pressure profile with, now, a gravity solely along y and a sloped mesh:

![image](https://github.com/user-attachments/assets/d4f0189a-62d1-4ed2-b719-972f2a24f7fd)

**Reply by wehrad** [reply]
I think I come to realize I get those results in FV because it may be stuck in a null-body solution or equivalent

This is the x-velocity solution with a slope of 10 degrees and periodic BCs on left/right for FE -and it matches the analytical solution-  (here with a 1e-7 sliding in x at the base but that doesn't matter): 
![image](https://github.com/user-attachments/assets/a70f239b-7641-4d4b-a045-b3bc1c18cd00)

This is the x-velocity solution with the same slope, periodic BCs based on your new object, in FV:
![Screenshot from 2025-01-02 14-45-09](https://github.com/user-attachments/assets/75c6c387-70ec-4545-92bf-2892427b36d7)

The ice is hardly moving in FV...! And I guess the nice vertical pressure profile constant along x in FE comes from the flow. Which is barely existent in my FV solution. 

Both simulations here are with the horizontal mesh and slanted gravity.

This may be due to FV struggling when there's no clear influx/outflux given?

**Reply by wehrad** [reply]
If I add some 1e-7 sliding in x at the base on the FV setup, it puts everything a bit more in motion but still, the flow induced by gravity should be ~ 5x higher than the sliding. Here it's driving everything...

![image](https://github.com/user-attachments/assets/6fde00ef-84e6-4036-b5e6-f3d07833c3a1)

More importantly, the sliding at the base should be of the same order at the side due to periodicity, but here it decreases..

it seems like the y velocity really shows the slope doesn't do anything, it's really similar to a flow on a flat surface, which shouldn't be the case. As shown in FE.

![image](https://github.com/user-attachments/assets/6523e5c8-0060-46f0-adb0-1224fadf5055)

**Reply by wehrad** [reply]
Now trying with periodic velocity on the inclined mesh (with the transform taking the slope of the mesh into account), I still get those near zero velocities with a extremely high viscosity (like 4-5 order of magnitudes above what it should be, which actually makes sense with the extremely low velocities)

**Reply by GiudGiud** [reply]
> and then for FV (on vel_x):

do you have the same kernel for vel_y active?

I think in terms of boundary conditions you should have:
- periodic for vel_x, achieved by setting a dirichletBC on one side equal to the velocity on the other side. (preferably with the BC being upwind, at the higher elevation)
- periodic for vel_y (not sure we need that, but maybe)
- periodic with an offset equal to rho g delta-h between the two sides. That offset is going to be driving the flow, because it's a pressure differential

**Reply by wehrad** [reply]
> do you have the same kernel for vel_y active?

Yes!

> - periodic with an offset equal to rho g delta-h between the two sides. That offset is going to be driving the flow, because it's a pressure differential

this makes me think that might be the one I don't have implemented correctly in FV for now, since pretty much no flow is being created. Interestingly, in FE, I don't have this offset either, but flow occurs and in the right way. This might show a difference between FE and FV? Where FV strictly needs this offset while FE generates it by itself?

**Reply by wehrad** [reply]
> this makes me think that might be the one I don't have implemented correctly in FV for now, since pretty much no flow is being created. Interestingly, in FE, I don't have this offset either, but flow occurs and in the right way. This might show a difference between FE and FV? Where FV strictly needs this offset while FE generates it by itself?

However this should be resulting in the inclined mesh case, but even with this inclined mesh, flow is not created. Which makes me even more puzzled.

**Reply by GiudGiud** [reply]
If you don't have this offset in finite element then something is wrong there. The pressure should not be periodic. 

Do you have this boundary condition on top of the new HydrostaticPressure ? This would not be correct

**Comment by wehrad** [comment]
It seemed like the issue I had with the periodic boundary was some sort of pressure null space, where no flow would be initiated. Now I made the parallel-sided slab very long with regards to thickness (10:1 ratio) and applied the overburden pressure (rho*g*H*sin(slope)) on both sides, which somehow kicked in flow so it could fully the true gravitational flow with no longitudinal pressure gradient. The slab seems flat here, but slope is applied by projecting gravity.

![image](https://github.com/user-attachments/assets/845b4e06-576c-4e11-8e83-f4fc5f2ae5d9)

![image](https://github.com/user-attachments/assets/73a57cf8-125c-44ba-9236-96771f74627a)

It gives a ~0.1-0.2% mismatch with the analytical solution on the vertical profile. That's on a very fine mesh though (1.6m resolution in depth for a depth of 100m), and gives quickly higher residuals for lower resolutions. I guess that's because the volume integration requires very high resolution close to the boundaries especially with strong gradients like the base here.

Does that make sense to you @GiudGiud ?

**Reply by GiudGiud** [reply]
Is there a periodic boundary condition applied here? On which variables?

**Reply by wehrad** [reply]
> Is there a periodic boundary condition applied here? On which variables?

No, I'm just applying the same `INSFVOutletPressureBC` on both ends with the ice overburden pressure

**Reply by wehrad** [reply]
I suppose the stress BC is just helping to kick the flow in. Then the actual flow within the ice slab is the actual gravitational flow I want to capture. The velocity profile near the center of the slab is independent of the pressure chosen (eg I did some tests with a simpler 1e5 on each side and got similar results).

**Reply by GiudGiud** [reply]
> I suppose the stress BC is just helping to kick the flow in.

can you do with an initial condition on velocity instead? 

A BC that remains acticve throughout the simulation will influence the result

**Reply by wehrad** [reply]
> can you do with an initial condition on velocity instead?

I tried having a random velocity field throughout the ice slab to start the flow but that didn't develop the gravitational flow I'm targeting... Or should I give the velocity field some structure with depth directly in the initial condition?

**Reply by GiudGiud** [reply]
it's ok to do a good initialization if you know the solution roughly. It will just save you computational time

**Reply by wehrad** [reply]
> can you do with an initial condition on velocity instead?

I tried both now while using the periodic BC, but somehow I can't get the gravitational flow to kick off with the periodic BC, only when I set an inlet and outlet pressure with a pseudo infinite slab...

With the periodic BC I still get the same results as some months ago (screenshot of discussion):
![image](https://github.com/user-attachments/assets/638ac850-0155-4ab3-a044-ed14788643ee)

**Reply by wehrad** [reply]
Here is a walkthrough of the steps I did. First, no BC on left nor right, only natural slip BC at the surface and no slip at the bottom. The pressure gives:

![image](https://github.com/user-attachments/assets/4104fde8-01d1-491d-a73d-8a449bf1050a)

Then I set the overburden pressure (rho*g*h*sin(slope)) at the right boundary:

![image](https://github.com/user-attachments/assets/c133f9bc-cad0-439b-9ecb-56296ac032c7)

And then only when I set the same overburden pressure both at the left and right boundary do I see the gravitational flow created without longitudinal pressure gradient:

![image](https://github.com/user-attachments/assets/13025114-4b23-41d9-9d8e-e73953a81a2a)

In that case the slab is not very long so boundary effects are very visible.

Here an example with a much longer slab to make in pseudo infinite:
![image](https://github.com/user-attachments/assets/9a7e92bf-8099-4186-bf73-489b3320e6ca)

**Reply by GiudGiud** [reply]
So in the much longer slab, should you see more of a slope/gradient in pressure like in the shorter case? The gravity / slope is the same right?

A problem with pressure BCs on both sides is well posed as long as there is pressure drop in the system, which you get from the no-slip BC at the bottom.

**Reply by wehrad** [reply]
> So in the much longer slab, should you see more of a slope/gradient in pressure like in the shorter case? The gravity / slope is the same right?

No the gradient in pressure should be the same, since velocity doesn't change along flow and the slope is constant. Which is what I observe, the pressure gradient is the same in both short and long cases.

> A problem with pressure BCs on both sides is well posed as long as there is pressure drop in the system, which you get from the no-slip BC at the bottom.

So you mean that pressure BCs on both sides would become an issue the moment there's a free slip at the bottom?

**Reply by GiudGiud** [reply]
> the pressure gradient is the same in both short and long cases.

ok great. It did not like it visually to me. 

> So you mean that pressure BCs on both sides would become an issue the moment there's a free slip at the bottom?

If you have free slip everywhere and laminar flow you dont have anything to create a pressure drop in the flow. So how would you get the pressure drop from one side to the other?
